lgc/util/llpcGfxRegHelper.h

```diff
@@ -0,0 +1,145 @@
+/*
+ ***********************************************************************************************************************
+ *
+ *  Copyright (c) 2020 Advanced Micro Devices, Inc. All Rights Reserved.
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+ /**
+  ***********************************************************************************************************************
+  * @file  llpcGfxRegHelper.h
+  * @brief LLPC header file: contains the definition of LLPC utility class GfxRegHelper.
+  ***********************************************************************************************************************
+  */
+
+#pragma once
+#include "llpcGfxRegHandler.h"
+#include "llpcTargetInfo.h"
+
+namespace lgc
+{
+
+using namespace llvm;
+
+// =====================================================================================================================
+// Struct maintains state for specific register
+// - pValue : Cache the value of the register which is being fetched already.
+// - isModified : Set to be true if the register is being settled with new value.
+struct BitsState
+{
+    Value* pValue = nullptr;
+    bool isModified = false;
+};
+
+// =====================================================================================================================
+// Helper class for handling graphics registers
+class GfxRegHelper : public GfxRegHandlerBase
+{
+protected:
+    GfxRegHelper(Builder* pBuilder, Value* pRegister);
+
+    virtual ~GfxRegHelper() = default;
+
+    // Get register
+    Value* GetRegCommon(uint32_t id);
```

Comment:
Where do the ids come from? What are the possible values for `id`?

Reply:
The id is the index of BitsInfo and BitsState
```
// =====================================================================================================================
// SqImgSampReg Bits infomation look up table (Gfx9-10)
// Refer to imported/chip/gfx9/gfx9_plus_merged_registers.h : SQ_IMG_SAMP_WORD
static constexpr BitsInfo g_sqImgSampRegBitsGfx9[static_cast<uint32_t>(SqSampRegs::SqSampRegsCount)] =
{
    { 0, 30, 2 }, // FilterMode
    { 2, 20, 2 }, // XyMagFilter
    { 2, 22, 2 }, // XyMinFilter
};

// SqImgSampRegisters Id
// Corresponding to g_sqImgSampRegBitsGfx9
enum class SqSampRegs
{
    FilterMode = 0,
    xyMagFilter,
    xyMinFilter,

    SqSampRegsCount,
};
```

Reply:
Acutally, we don't expect user to use `GfxRegHelper` directly, but only implement its own register helper class, such as `SqImgSampRegHelper`

Reply:
Sounds good to me. Can we have it in the comments?

Reply:
Updated.